### PR TITLE
Fix last_event_id() in nested scopes

### LINKED ISSUE #3113: last_event_id() crash: AttributeError: 'Scope' object has no attribute '_last_event_id'.
### How do you use Sentry?

Sentry Saas (sentry.io)

### Version

master

### Steps to Reproduce

Run this file:

```python
import sentry_sdk

sentry_sdk.init()

with sentry_sdk.push_scope() as scope:
    sentry_sdk.last_event_id()
```

### Expected Result

No crash.

### Actual Result

Crash:

```
$ python example.py
Traceback (most recent call last):
  File "/.../example.py", line 6, in <module>
    sentry_sdk.last_event_id()
  File "/.../sentry_sdk/api.py", line 343, in last_event_id
    return Scope.last_event_id()
           ^^^^^^^^^^^^^^^^^^^^^
  File "/.../sentry_sdk/scope.py", line 330, in last_event_id
    return cls.get_isolation_scope()._last_event_id
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'Scope' object has no attribute '_last_event_id'. Did you mean: 'last_event_id'?
```

#3064 added `last_event_id()` and `_last_event_id` but I think it missed copying `_last_event_id` in `Scope.__copy__`.

### PR DESCRIPTION
**Pull request by adamchainz**
Fixes #3113.

I added two tests, one that generally checks `Scope.__copy__()` copies everything that it should, and one top-level one that reproduces the bug seen. On `master`, both fail with:

```
AttributeError: 'Scope' object has no attribute '_last_event_id'. Did you mean: 'last_event_id'?
```

@szokeasaurusrex please take a look :)

### CODE REVIEW

File: sentry_sdk/scope.py [hunk 1]
```diff
@@ -208,9 +208,6 @@ def __init__(self, ty=None, client=None):
         incoming_trace_information = self._load_trace_data_from_env()
         self.generate_propagation_context(incoming_data=incoming_trace_information)
 
-        # self._last_event_id is only applicable to isolation scopes
-        self._last_event_id = None  # type: Optional[str]
-
```
Comment (szokeasaurusrex) [inline sentry_sdk/scope.py:213]: While I agree that it likely makes sense to move these lines to the `clear` function, I believe this change should be considered in a separate PR, since it does not appear to be needed to fix #3113; the tests you added pass even if I undo this change.

File: sentry_sdk/scope.py [hunk 2]
```diff
@@ -677,6 +676,9 @@ def clear(self):
 
         self._propagation_context = None
 
+        # self._last_event_id is only applicable to isolation scopes
+        self._last_event_id = None  # type: Optional[str]
+
```
Comment (szokeasaurusrex) [inline sentry_sdk/scope.py:681]: Likely belongs in separate PR, see previous comment.

File: tests/test_scope.py [hunk 3]
```diff
@@ -25,6 +25,10 @@ def test_copying():
     s1.set_tag("foo", "bar")
 
     s2 = copy.copy(s1)
+    # Check all attributes are copied
+    for name in set(Scope.__slots__) - {"client"}:
+        getattr(s2, name)
+
```
Comment (szokeasaurusrex) [inline tests/test_scope.py:31]: I would prefer to add this check to a separate unit test, since it is somewhat unrelated to the tag copying checks that this existing test is checking. The comment should also be updated to something like "Check that all attributes, except `client`, are present on the new scope", as the check only verifies the attributes' presence, and does not check whether the values were actually copied.

**Comment by sl0thentr0py** [comment]
@antonpirker assigning this to you since I'm not sure what the desired semantics here should be (since it only should be on the isolation scope and so on..)

File: tests/test_basics.py [hunk 4]
```diff
@@ -800,3 +800,11 @@ def test_last_event_id_transaction(sentry_init):
         pass
 
     assert last_event_id() is None, "Transaction should not set last_event_id"
+
+
+def test_last_event_id_scope(sentry_init):
+    sentry_init(enable_tracing=True)
+
+    # Should not crash
+    with push_scope() as scope:
```
Comment (szokeasaurusrex) [inline tests/test_basics.py:809]:
```suggestion
    with isolation_scope() as scope:
```

`push_scope` is [deprecated](https://github.com/getsentry/sentry-python/blob/master/CHANGELOG.md#deprecated) since version 2.0.0, so let's use the new `isolation_scope` function here, instead. Using `isolation_scope` also fails on master but passes on this branch.

**Review by szokeasaurusrex — COMMENTED** [review]
Thank you for the contribution! I have requested a few changes, please see my comments.

**Comment by szokeasaurusrex** [comment]
@sl0thentr0py, removing @antonpirker's assignment since I have already reviewed – I wrote the original Scope-based `last_event_id` implementation (#3064), so I am familiar with this code

**Comment by szokeasaurusrex** [comment]
Also, I split your changes to clear `_last_event_id` in the `clear` method into #3124

**Comment by adamchainz** [comment]
Thank you for the fix. I didn’t have to look at this because I wasn’t working on the related contract.
